Source/WebCore/dom/FullscreenManager.cpp

```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
 
     INFO_LOG(LOGIDENTIFIER);
 
-    // To achieve that aim, remove all the elements from the top document's stack except for the first before
-    // calling webkitExitFullscreen():
-    RefPtr fullscrenElement = topDocument.fullscreenManager().fullscreenElement();
-    topDocument.fullscreenManager().m_fullscreenElementStack = { WTFMove(fullscrenElement) };
+    m_pendingExitFullscreen = true;
 
-    topDocument.fullscreenManager().exitFullscreen();
+    m_document.eventLoop().queueTask(TaskSource::MediaElement, [this, &topDocument, identifier = LOGIDENTIFIER] {
+        if (!topDocument.page()) {
+            INFO_LOG(LOGIDENTIFIER, "Top document has no page.");
+            return;
+        }
+
+        // This triggers finishExitFullscreen with ExitMode::Resize, which fully exits the document.
+        if (auto* fullscreenElement = topDocument.fullscreenManager().fullscreenElement())
+            page()->chrome().client().exitFullScreenForElement(fullscreenElement);
+        else
+            INFO_LOG(LOGIDENTIFIER, "Top document has no fullscreen element");
+    });
 }
 
-void FullscreenManager::exitFullscreen()
+static Vector<RefPtr<Document>> documentsToUnfullscreen(Document& doc)
 {
-    // The exitFullscreen() method must run these steps:
-
-    // 1. Let doc be the context object. (i.e. "this")
-    Document* currentDoc = &document();
-
-    // 2. If doc's fullscreen element stack is empty, terminate these steps.
-    if (m_fullscreenElementStack.isEmpty()) {
-        // If there is a pending fullscreen element but an empty fullscreen element stack,
-        // there is a pending task in requestFullscreenForElement(). Cause it to cancel and fire an error
-        // by clearing the pending fullscreen element.
-        INFO_LOG(LOGIDENTIFIER, "Cancelling pending fullscreen request.");
-        m_pendingFullscreenElement = nullptr;
-        return;
+    Vector<RefPtr<Document>> docs { RefPtr { &doc } };
+    while (true) {
+        auto lastDoc = docs.last();
+        if (!lastDoc)
+            continue;
+        ASSERT(lastDoc->fullscreenManager().fullscreenElement());
+        if (!lastDoc->fullscreenManager().isSimpleFullscreenDocument())
+            break;
+        auto parentFrame = lastDoc->frame()->ownerElement();
```

Comment:
What guarantees that `lastDocument->frame()` is non-null here?

A local variable that is an element should probably not be named frame. Maybe `frameOwner` instead of `parentFrame`?